### ISSUE TITLE
Server and CLI don' t exclude the same directories

### ISSUE THREAD

**Issue by EricDepagne** [astral-sh/ruff#11587]
I' m using ruff server in vscode.
Ruff version: 0.4.5
VSCode version : 1.89.1 (Universal)

My ruff.toml contains the following exclusion:
exclude = ["\*/tests/\*"]

I run ruff in a terminal at the same time as the ruff server in vscode.
The intersting part of the output of ruff -w is:

```
[23:26:28 PM] Starting linter in watch mode...

warning: Unexpected `# ruff: noqa` directive at conftest.py:10. File-level suppression comments must appear on their own line. For line-level suppression, omit the `ruff:` prefix.
warning: Unexpected `# ruff: noqa` directive at bikeanalysis/activities/models/activities.py:228. File-level suppression comments must appear on their own line. For line-level suppression, omit the `ruff:` prefix.
[23:26:28 PM] Found 19 errors. Watching for file changes.
```
Only 19 errors found

In vscode, the same errors are flagged, as expected, but I have also all the warnings for all my test files, that are in one of the tests/ directories.

For instance, this one:
```

def test_missing_all_streams():
    streams = []
    assert len(get_missing_streams(streams=streams)) == 11
```
triggers D103.

The ruff that runs in my terminal does not show any errors for files in my tests/ directory as expected from the configuration.

Am I missing something when configuring ruff server in VSCode?

**Comment by charliermarsh** [comment]
Might be a real bug in the new server. I'll try to repro tonight.

**Comment by charliermarsh** [comment]
`exclude = ["*/tests/*"]` doesn't actually work for me, even on the command-line. `exclude = ["**/tests/*"]` does work though.

**Comment by charliermarsh** [comment]
I can reproduce with `exclude = ["**/tests/*"]` though.

**Comment by charliermarsh** [comment]
I think `ruff server` just doesn't respect exclusions yet if you open an excluded file directly.

**Comment by messense** [comment]
#11590 breaks my format workflow, I'd expect the vscode `format document` and `format cell` commands to format the python/notebook file even if it's excluded in configuration file, because I'm explicitly asking vscode to format it.

BTW, `ruff format <path to file>` works regardless of file exclusions config, so IMO manually call vscode `format document` should have the same behavior.

**Comment by MichaReiser** [comment]
> https://github.com/astral-sh/ruff/pull/11590 breaks my format workflow, I'd expect the vscode format document and format cell commands to format the python/notebook file even if it's excluded in configuration file, because I'm explicitly asking vscode to format it.

Hmm, I do understand the desire that explicitly issuing format document or format cells should format the code even if it is excluded. The challenge we face is that the LSP request, as far as I know, contains no information whether the request was triggered implicitly (by enabling format on save) or explicitly (by issuing the command). Implicit request should respect the ignores, I'm open to change that explicit requests don't.

**Comment by messense** [comment]
An new option would be nice for people don't use `format on save`.

**Comment by dhruvmanila** [comment]
> The challenge we face is that the LSP request, as far as I know, contains no information whether the request was triggered implicitly (by enabling format on save) or explicitly (by issuing the command)

Yeah, this is correct. The server has no way of knowing whether the request was triggered automatically or manually.

It seems reasonable to add a new option and is related to https://github.com/astral-sh/ruff/issues/12176. @messense Is it that the extension should format / lint _every_ document that's opened in the editor, thus not considering the includes / excludes ?

**Comment by masaccio** [comment]
Is this really about files that are opened explicitly? I'm finding that VScode flags warnings from files that are not open but are in my workspace. In my case these files:

<img width="267" alt="Screenshot 2024-10-14 at 18 15 23" src="https://github.com/user-attachments/assets/73f904dd-7f0d-475f-9c33-02f17dc41c2a">

Despite:

``` tool
[tool.ruff]
exclude = [
  # Machine-generated files
  "**/.bootstrap/*",
  "**/.tox/*",
  "**/.vscode/*",
  "**/src/numbers_parser/generated/*",
]
```

Whether they're included by the server is a bit random. Is there a VScode log that would be helpful?

**Comment by dhruvmanila** [comment]
> Is this really about files that are opened explicitly? I'm finding that VScode flags warnings from files that are not open but are in my workspace.

Can you say more about this? Is it that the diagnostics window includes violations from files that aren't opened in the editor? Are those diagnostics coming from Ruff because we don't support workspace wide diagnostics? Do you have `python.analysis.diagnosticMode` set to `workspace`? That setting is used by the Python extension to include workspace wide diagnostics.
